### PR TITLE
feat(core): add support for optional dependencies

### PR DESCRIPTION
**Pull request by fahrradflucht**

### CODE REVIEW

File: packages/core/src/feature-service-registry.ts [hunk 1]
```diff
@@ -177,26 +182,31 @@ export class FeatureServiceRegistry implements FeatureServiceRegistryLike {
     const bindings = new Map<string, FeatureServiceBinding<unknown>>();
     const featureServices: FeatureServices = Object.create(null);
```
Comment (clebert) [inline packages/core/src/feature-service-registry.ts]: please remove linebrak

File: packages/core/src/feature-service-registry.ts [hunk 2]
```diff
@@ -177,26 +182,31 @@ export class FeatureServiceRegistry implements FeatureServiceRegistryLike {
...
+      const binding = this.bindFeatureService(
+        providerId,
+        consumerUid,
+        allDependencies[providerId],
+        {optional: !dependencies.hasOwnProperty(providerId)}
+      );
 
-        featureServices[providerId] = binding.featureService;
+      if (!binding) {
+        break;
```
Comment (clebert) [inline packages/core/src/feature-service-registry.ts]: please write first a test that fails and then change this statement to continue

File: packages/core/src/__tests__/feature-service-registry.test.ts [hunk 3]
```diff
@@ -352,25 +352,74 @@ describe('FeatureServiceRegistry', () => {
       });
     });
```
Comment (clebert) [inline packages/core/src/__tests__/feature-service-registry.test.ts]: please test the scenario that an optional dependency mismatch does not fail

File: packages/core/src/__tests__/feature-service-registry.test.ts [hunk 4]
```diff
@@ -352,25 +352,74 @@ describe('FeatureServiceRegistry', () => {
...
+            registry.bindFeatureServices({
+              id: 'foo',
+              optionalDependencies: {b: '1.0', a: '1.1'}
+            })
+          ).toEqual({
+            featureServices: {a: featureServiceA},
+            unbind: expect.any(Function)
+          });
+
+          expect(binderA.mock.calls).toEqual([['foo']]);
```
Comment (clebert) [inline packages/core/src/__tests__/feature-service-registry.test.ts]: test b

File: packages/core/src/__tests__/feature-service-registry.test.ts [hunk 5]
```diff
@@ -352,25 +352,74 @@ describe('FeatureServiceRegistry', () => {
...
+            registry.bindFeatureServices({
+              id: 'foo',
+              optionalDependencies: {a: '1.1', b: '^1.0'}
+            })
+          ).toEqual({
+            featureServices: {a: featureServiceA, b: featureServiceB},
+            unbind: expect.any(Function)
+          });
+
+          expect(binderA.mock.calls).toEqual([['b'], ['foo']]);
```
Comment (clebert) [inline packages/core/src/__tests__/feature-service-registry.test.ts]: binderB

File: packages/core/src/__tests__/feature-service-registry.test.ts (hunk 4 above)
Comment (fahrradflucht) [reply packages/core/src/__tests__/feature-service-registry.test.ts]: B is not even registered. It is impossible that `binderB` does not exist in this test context.

**Pull request merged**